src/gpu/nvidia/cudnn_convolution.cpp

```diff
@@ -138,6 +132,8 @@ status_t cudnn_convolution_bwd_data_t::execute_convolution(
             filter_scratch_acc
                     = std::make_shared<scratch_acc_t>(CTX_SCRATCH_ACCESSOR(
                             memory_tracking::names::key_conv_cudnn_filter));
+            filter_offset = CTX_SCRATCH_OFFSET(
```

Comment:
From what I understand the issue affects three scenarios:
1. When there are more than 1 scratchpad booked by the library
2. When a user passes sub-buffers as scratchpad
3. When a user passes sub-buffers as memory arguments

The first scenario can be handled by adding the offsets explicitly. The other two require a SYCL API to query offsets that were used for creating a sub-buffer. From what I see SYCL doesn't specify any way to extract those offsets. 

Ideally, we need the API so that we can simply adjust the pointer in `inline T memory(const compat::interop_handle &ih, U acc)`.

SYCL 2020 specification is vague about sub-buffers. At the first glance it seems that returning the base pointer is the correct behavior: 
> A sub-buffer object can be created which is a sub-range reference to a base buffer. This sub-buffer can be used to create accessors to the base buffer, which have access to the range specified at time of construction of the sub-buffer.

Reply:
I'm concerned that the proposed way of handling the 1 scenario is error prone and we can easily run into a similar issue in other primitives.
Would that be possible to adjust `memory(...)` function so that it takes buffer memory storage to extract the base offset and apply it?
```cpp
template <typename T, typename U>
inline T memory(const compat::interop_handle &ih, U acc, const sycl_buffer_memory_storage_t *buffer_storage) {
    auto *ptr = compat::get_native_mem<uint8_t *>(ih, acc) + buffer_storage->base_offset();
    return reinterpret_cast<T>(ptr); 
} 
```

Reply:
Per feedback from the compiler and specification teams the behavior looks incorrect. The behavior is being clarified [here](https://github.com/KhronosGroup/SYCL-Docs/pull/197/files#r787930740).

Reply:
> I'm concerned that the proposed way of handling the 1st scenario is error prone and we can easily run into a similar issue in other primitives. Would that be possible to adjust `memory(...)` function so that it takes buffer memory storage to extract the base offset and apply it?
> 
> ```c++
> template <typename T, typename U>
> inline T memory(const compat::interop_handle &ih, U acc, const sycl_buffer_memory_storage_t *buffer_storage) {
>     auto *ptr = compat::get_native_mem<uint8_t *>(ih, acc) + buffer_storage->base_offset();
>     return reinterpret_cast<T>(ptr); 
> } 
> ```
> 
> Meanwhile I started a discussion about the problem with the compiler team.

Is there a benefit in using the entire buffer_storage object instead of just requesting the offset?

Reply:
You mean passing the requested offset instead of storage? If so, I think there is no difference.

Reply:
Yeah, that's what I meant. There have been issues with capturing the buffer_storage object because it's passed as a unique_ptr from `get_memory_storage` in the scratchpad grantor. We've tried multiple different approaches but ultimately each time the destructor for the object is called by the unique_ptr before the host_task can be completed. Do you have any suggestions around this?